src/backend/distributed/master/worker_node_manager.c

```diff
@@ -289,118 +276,59 @@ WorkerNode *
 
-	workerNode = (WorkerNode *) hash_seq_search(&status);
-	while (workerNode != NULL)
+	while ((workerNode = hash_seq_search(&status)) != NULL)
 	{
-		if (workerNode->inWorkerFile)
+		int nameCompare = strncmp(workerNode->workerName, hostname, WORKER_LENGTH);
+		if (nameCompare == 0)
 		{
-			int nameCompare = strncmp(workerNode->workerName, hostname, WORKER_LENGTH);
-			if (nameCompare == 0)
-			{
-				hash_seq_term(&status);
-				break;
-			}
+			/* we need to terminate the scan since we break */
+			hash_seq_term(&status);
+			break;
 		}
-
-		workerNode = (WorkerNode *) hash_seq_search(&status);
 	}
 
 	return workerNode;
 }
 
 
-/* Returns the number of live nodes in the cluster. */
+/*
+ * WorkerGetLiveNodeCount returns the number of live nodes in the cluster.
+ * */
 uint32
 WorkerGetLiveNodeCount(void)
 {
-	WorkerNode *workerNode = NULL;
-	uint32 liveWorkerCount = 0;
-
-	HASH_SEQ_STATUS status;
-	hash_seq_init(&status, WorkerNodesHash);
-
-	workerNode = (WorkerNode *) hash_seq_search(&status);
-	while (workerNode != NULL)
-	{
-		if (workerNode->inWorkerFile)
-		{
-			liveWorkerCount++;
-		}
-
-		workerNode = (WorkerNode *) hash_seq_search(&status);
-	}
+	HTAB *workerNodeHash = GetWorkerNodeHash();
+	uint32 liveWorkerCount = hash_get_num_entries(workerNodeHash);
 
 	return liveWorkerCount;
 }
 
 
-/* Inserts the live worker nodes to a list, and returns the list. */
+/*
+ * WorkerNodeList iterates over the hash table that includes the worker nodes, and adds
+ * them to a list which is returned.
+ */
 List *
 WorkerNodeList(void)
 {
 	List *workerNodeList = NIL;
 	WorkerNode *workerNode = NULL;
-
+	HTAB *workerNodeHash = GetWorkerNodeHash();
 	HASH_SEQ_STATUS status;
-	hash_seq_init(&status, WorkerNodesHash);
-
-	workerNode = (WorkerNode *) hash_seq_search(&status);
-	while (workerNode != NULL)
-	{
-		if (workerNode->inWorkerFile)
-		{
-			workerNodeList = lappend(workerNodeList, workerNode);
-		}
-
-		workerNode = (WorkerNode *) hash_seq_search(&status);
-	}
-
-	return workerNodeList;
-}
 
+	hash_seq_init(&status, workerNodeHash);
 
-/*
- * WorkerNodeActive looks up a worker node with the given name and port number
- * in the current membership list. If such a worker node exists, the function
- * returns true.
- */
-bool
-WorkerNodeActive(const char *nodeName, uint32 nodePort)
-{
-	bool workerNodeActive = false;
-	bool handleFound = false;
-	WorkerNode *workerNode = NULL;
-	void *hashKey = NULL;
-
-	WorkerNode *searchedNode = (WorkerNode *) palloc0(sizeof(WorkerNode));
-	strlcpy(searchedNode->workerName, nodeName, WORKER_LENGTH);
-	searchedNode->workerPort = nodePort;
-
-	hashKey = (void *) searchedNode;
-	workerNode = (WorkerNode *) hash_search(WorkerNodesHash, hashKey,
-											HASH_FIND, &handleFound);
-	if (workerNode != NULL)
+	while ((workerNode = hash_seq_search(&status)) != NULL)
 	{
-		if (workerNode->inWorkerFile)
-		{
-			workerNodeActive = true;
-		}
+		workerNodeList = lappend(workerNodeList, workerNode);
```

Comment:
I think only remaining question is whether we should copy workerNode here. Since workerNode is allocated in the hash and the hash could be destroyed whenever a subsequent call to GetWorkerNodeHash occurs, the WorkerNode is only safe to use until the next call to GetWorkerNodeHash. If we allocate new memory and copy it in there, the caller is free to use it.

I think we should either make a copy here or document that the returned WorkerNode is only safe until the next call to GetWorkerNodeHash.